File: daily/2024/592_Fraction_Addition_and_Subtraction.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    int gcd(int a, int b){
        if (a < b) swap(a, b);
        if (b == 0) return a;
        return gcd(b, a % b);
    }
    string fractionAddition(string expression) {
        int index = 0;
        int s = expression.length();
        int Numerator = 0;
        int Denominator = 1;
        while(index < s){
            int signal = 1;
            if (expression[index] == '-') signal = -1;
            if (!isdigit(expression[index])) index++;
            int temp_Numerator = 0;
            int temp_Denominator = 0;
            while(index < s && isdigit(expression[index])) temp_Numerator = temp_Numerator * 10 + (expression[index++] - '0');
            index++;
            while(index < s && isdigit(expression[index])) temp_Denominator = temp_Denominator * 10 + (expression[index++] - '0');
            // cout << "accu = " << ((Numerator >= 0)? '+':'-') << abs(Numerator) << " / " << Denominator << endl;
            // cout << "temp = " << ((signal == 1)? '+':'-') << temp_Numerator << " / " << temp_Denominator << endl;
            int common_factor = gcd(Denominator, temp_Denominator);
            // cout << "common factor = " << common_factor << endl;
            Numerator = Numerator * (temp_Denominator / common_factor) + temp_Numerator * (Denominator / common_factor) * signal;
            // cout << "Numerator = " << Numerator << endl;
            if (Numerator == 0) Denominator = 1;
            else Denominator *= (temp_Denominator / common_factor);
            // cout << "Denominator = " << Denominator << endl;
            common_factor = gcd(abs(Numerator), Denominator);
            // cout << "common factor = " << common_factor << endl;
            Numerator /= common_factor;
            Denominator /= common_factor;
        }
        string ans;
        ans = to_string(Numerator) + '/' + to_string(Denominator);
        return ans;
    }
};

int main(){
    string expression = "-1/2+1/2";
    // string expression = "-1/2+1/2+1/3";
    // string expression = "1/3-1/2";
    // string expression = "-4/7-3/4+2/3";
    Solution S;
    cout << S.fractionAddition(expression) << endl;
    return 0;
}

File: daily_python/592_Fraction_Addition_and_Subtraction.py
import os
import sys

class Solution:
    def gcd(self, a, b)-> int:
        if b == 0:
            return a
        return self.gcd(b, a % b)
    
    def fractionAddition(self, expression: str) -> str:
        if expression[0] != '-':
            expression = '+' + expression
        index = 0
        s = len(expression)
        numerator = 0
        denominator = 1
        while(index < s):
            signal = 1 if expression[index] == '+' else -1
            index += 1
            curr_numerator = 0
            while(index < s and expression[index].isdigit()):
                curr_numerator = curr_numerator * 10 + int(expression[index])
                index += 1
            curr_numerator *= signal
            index += 1
            curr_denominator = 0
            while(index < s and expression[index].isdigit()):
                curr_denominator = curr_denominator * 10 + int(expression[index])
                index += 1
            # print(f'{curr_numerator} / {curr_denominator}')
            numerator = numerator * curr_denominator + curr_numerator * denominator
            denominator *= curr_denominator
            common_factor = self.gcd(numerator, denominator)
            numerator //= common_factor
            denominator //= common_factor
        return str(numerator) + '/' + str(denominator)
        
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            expression = lines[0].split('=')[1].strip()[1:-2]
            dataset.append(expression)
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for expression in dataset:
        results.append(solution.fractionAddition(expression))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')